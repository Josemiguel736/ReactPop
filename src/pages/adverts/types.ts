interface User {
    name: string;
    username: string;
  }
  
  export interface Advert {
    id: number;
    userId: number;
    content: string;
    updatedAt: string;
    user: User;
  }
  
  export interface AdvertContent {
    content: string;
  }
  