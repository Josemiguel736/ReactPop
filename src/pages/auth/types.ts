export interface Credentials {
	email: string;
	password: string;
}

export interface Login {
	accessToken: string;
}

export interface Me {
	id: string;
	createdAt: string;
	email: string;
	username: string;
	name: string;
}
