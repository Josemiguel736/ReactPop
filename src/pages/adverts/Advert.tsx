import { AdvertType } from "./types";
import imageNotFound from "../../assets/imageNotFound.jpg"

interface Props {
  advert: AdvertType;
}

const Advert = ({ advert }: Props) => {
  const { createdAt, name, sale, price, tags } = advert;
  const date = new Date(createdAt)

  return (
    <div className="bg-sky-700 text-amber-50 p-4 flex flex-col items-center text-center rounded-xl shadow-md">
  <h3 className="text-3xl font-bold mb-3">{name}</h3>

  <div className="w-full h-48 bg-gray-200 overflow-hidden rounded-lg">
    <img
      className="w-full h-full object-cover"
      src={advert.photo ? advert.photo : imageNotFound}
      alt={`Imagen del producto ${name}`}
    />
  </div>

  <span className="mt-3">{sale}</span>

  <div className="mb-3">
    <span className="text-[20px] mr-2.5">Precio: <strong>{price}</strong></span>
    <span>{advert.sale ? "Compra" : "Venta"}</span>
  </div>

  <div className="flex flex-col text-center justify-center bg-sky-400 pl-5 pr-5 rounded-lg">
    <span>Tags</span>
    <span className="text-[10px]">{tags}</span>
  </div>

  <span className="mt-2">Publicado el {date.toLocaleDateString("es-ES")}</span>
</div>

  );
};

export default Advert;
