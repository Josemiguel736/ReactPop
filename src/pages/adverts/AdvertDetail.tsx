import { useEffect, useState } from 'react';
import { useNavigate, useParams } from 'react-router-dom';
import { AdvertType } from './types';
import { deleteAdvert, getAdvert } from './service';
import imageNotFound from '../../assets/imageNotFound.jpg';
import Button from '../../components/shared/Button';
import ConfirmLogout from '../../components/shared/ConfirmButton';
import { ApiClientError } from '../../api/error';
import { isApiClientError } from '../../api/client';
import ErrorSpan from '../../components/errors/ErrorSpan';
import LoadingPage from '../../components/shared/loadingPage/LoadingPage';
import Page500 from '../ErrorPages/500';

function AdvertDetail() {
	const params = useParams();
	const navigate = useNavigate();

	const [isLoading, setIsLoading] = useState(true);
	const [advert, setAdvert] = useState<AdvertType | null>(null);
	const [error, setError] = useState<ApiClientError | null>(null);

	useEffect(() => {
		const advertParams = async () => {
			try {
				if (params.advertId) {
					const advertInfo = await getAdvert(params.advertId); // petición a la API para obtener el anuncio individual
					setAdvert(advertInfo);
				}
			} catch (error) {
				if (isApiClientError(error)) {
					if (error.code === 'NOT_FOUND') {
						console.log('ERROR 404');
						// Si hay un error en la petición a la API  o un error genérico redirigimos a la página de Not Found
						navigate('/404');
					} else {
						console.warn(
							'ERROR IN API CALL TO ADVERT DETAIL FROM ADVERT DETAIL',
							error,
						);
					}
				} else if (error instanceof Error) {
					console.warn('GENERIC ERROR IN ADVERT DETAILS', error);
					navigate('/404');
				}
			} finally {
				setIsLoading(false); // Manejo del estado de carga
			}
		};
		advertParams();
	}, [params.advertId, navigate]);

	const [isClicked, setIsClicked] = useState(false);

	const handleSubmit = async () => {
		if (advert) {
			try {
				setIsLoading(true);
				await deleteAdvert(advert.id);
				navigate('/');
			} catch (error) {
				if (isApiClientError(error)) {
					setError(error);
					console.warn(
						'ERROR IN API CALL TO DELETE ADVERT FROM ADVERT DETAIL',
						error,
					);
				} else if (error instanceof Error) {
					console.warn('GENERIC ERROR IN ADVERT DETAILS', error);
					navigate('/404');
				}
			} finally {
				setIsLoading(false);
			}
		}
	};

	return isLoading ? (
		<LoadingPage />
	) : advert ? (
		<div className="mt-3 flex flex-col items-center justify-center  ">
			<div className="bg-sky-700 text-amber-50 p-6 flex flex-col items-center text-center rounded-xl shadow-lg ">
				<img
					className="w-full max-h-72 object-cover rounded-lg mb-4"
					src={advert.photo || imageNotFound}
					alt={`Imagen de ${advert.name}`}
				/>

				<h3 className="text-2xl font-bold mb-2">Producto: {advert.name}</h3>
				<span className="mb-1">{advert.sale ? 'En venta' : 'Se busca'}</span>
				<span className="mb-3">Precio: {advert.price}€</span>

				<div className="flex flex-col items-center justify-center bg-sky-400 p-2 rounded-md mb-3 w-full">
					<span className="font-semibold">Tags</span>
					<div className="flex flex-row">
						{advert.tags.map((tag) => (
							<span key={tag} className="mr-1.5">
								{tag}
							</span>
						))}
					</div>
				</div>

				<span className="text-sm mb-4">
					Publicado el {new Date(advert.createdAt).toLocaleDateString('es-ES')}
				</span>

				{isClicked ? (
					<div className="flex flex-col items-center gap-2 ">
						<p className="mb-2">
							¿Estás seguro de que quieres borrar el producto?
						</p>
						<ConfirmLogout
							titlePrimary="Cancelar"
							titleSecondary="Borrar Producto"
							setIsClicked={setIsClicked}
							handleSubmit={handleSubmit}
						/>
					</div>
				) : (
					<Button
						onClick={() => setIsClicked(true)}
						$variant="primary"
						className="mt-4"
					>
						Borrar Producto
					</Button>
				)}

				{error instanceof ApiClientError && (
					<ErrorSpan>
						Ha sucedido un error al borrar el producto, por favor inténtalo de
						nuevo más tarde
					</ErrorSpan>
				)}
			</div>
		</div>
	) : (
		<Page500 error={new Error('Hemos tenido un problema')} />
	);
}

export default AdvertDetail;
