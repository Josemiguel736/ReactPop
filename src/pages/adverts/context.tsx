import { createContext, useContext } from 'react';
import { AdvertType } from './types';

interface FilterContextProps {
	filteredAdverts: AdvertType[] | null;
	setFilteredAdverts: (adverts: AdvertType[]) => void;
}

export const FilterContext = createContext<FilterContextProps>({
	filteredAdverts: null,
	setFilteredAdverts: () => undefined,
});

export function useFilter() {
	const filtertContext = useContext(FilterContext);
	return filtertContext;
}
