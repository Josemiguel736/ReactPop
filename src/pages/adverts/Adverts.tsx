import { Link } from 'react-router-dom';
import Advert from './Advert';
import { AdvertType } from './types';
import EmptyList from './EmpltyList';
import { useState } from 'react';
import { useFilter } from './context';

interface Props {
	adverts: AdvertType[];
}
function Adverts({ adverts }: Props) {
	const { filteredAdverts } = useFilter();

	const adds = filteredAdverts ?? adverts;

	const skip = 6;

	const [position, setPosition] = useState<number>(1);

	const [paginatedFilter, setPaginate] = useState<number[]>([0, skip]);

	const limit = Math.ceil(adds.length / skip);

	const handlePaginate = (isNext: boolean) => {
		// l칩gica de la paginaci칩n
		let newPosition = isNext ? position + 1 : position - 1;
		newPosition = Math.max(1, Math.min(newPosition, limit));
		setPosition(newPosition);
		setPaginate([skip * (newPosition - 1), skip * newPosition]);
	};

	if (position > limit) {
		setPosition(1);
		setPaginate([0, skip]);
	}

	const paginatedAdds =
		adds.slice(paginatedFilter[0], paginatedFilter[1]) ?? []; // crear un nuevo array con los anuncios que se van a mostrar en la p치gina

	return (
		<div className="container mx-auto ml-4">
			{paginatedAdds.length ? (
				<article className="flex justify-center items-center">
					<ul className="grid grid-cols-1 lg:grid-cols-3 gap-6 w-full">
						{paginatedAdds.map((advert) => (
							<li key={advert.id} className="mt-3.5">
								<Link to={`/adverts/${advert.id}`}>
									<Advert advert={advert} />
								</Link>
							</li>
						))}
					</ul>
				</article>
			) : (
				<EmptyList />
			)}
			<div className="flex justify-center items-center gap-4 mt-1.5 ">
				<button
					onClick={() => handlePaginate(false)}
					className="bg-blue-600 text-white px-6 py-3  shadow-md hover:bg-blue-700 transition-all disabled:bg-gray-400 "
					disabled={position <= 1}
				>
					Atr치s
				</button>

				<button
					onClick={() => handlePaginate(true)}
					className="bg-green-500 text-white px-6 py-3  shadow-md hover:bg-green-600 transition-all disabled:bg-gray-400 "
					disabled={position >= limit}
				>
					Siguiente
				</button>
			</div>
		</div>
	);
}

export default Adverts;
