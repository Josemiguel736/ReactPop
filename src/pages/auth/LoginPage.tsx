import { useEffect, useState } from "react";
import FormField from "../../components/shared/FormField";
import Button from "../../components/shared/Button";
import { login } from "./service";
import { useLocation, useNavigate } from "react-router-dom";
import { useAuth } from "./context";
import { validateEmail } from "../../utils/validate";
import { ApiClientError } from "../../api/error";
import { isApiClientError } from "../../api/client";
import ErrorSpan from "../../components/errors/ErrorSpan";

function LoginPage() {
  const { onLogin } = useAuth();

  const location = useLocation();

  const navigate = useNavigate();

  const [error, setError] = useState<ApiClientError | null>(null);

  const [isLoading, setIsLoading] = useState(false);

  const [checked, setIsChecked] = useState(false);

  const [emailIsValid, validEmail] = useState(false);

  const [credentials, setCredentials] = useState({
    email: "",
    password: "",
  });

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setCredentials((credentials) => ({
      ...credentials,
      [event.target.name]: event.target.value,
    }));
  };

  const handleValidateEmail = () => {
    const isValid = validateEmail(email);
    validEmail(isValid);
  };

  const { email, password } = credentials;

  useEffect(() => {
    handleValidateEmail();
  }, [email]);

  const isDisabled = !email || !password || isLoading || !emailIsValid;

  const handleSubmit = async (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    try {
      setIsLoading(true);
      await login(credentials, checked);
      onLogin();
      const to = location.state?.from ?? "/";
      navigate(to, { replace: true });
    } catch (error) {
      if (isApiClientError(error)) {
        setError(error);
      }
      console.warn(error);
    } finally {
      setIsLoading(false);
    }
  };

  return (
    <div className="h-1/3 mt-10 flex items-center justify-center">
      <form
        onSubmit={handleSubmit}
        className="text-white p-6 rounded-lg shadow-lg w-1/2 bg-sky-700 flex flex-col justify-center items-center text-center"
      >
        <h1 className="text-2xl mt-2.5 mb-1">Iniciar Sesión</h1>

        <FormField
          type="text"
          name="email"
          value={credentials.email}
          onChange={handleChange}
          className="mt-4 border-2 rounded-lg "
          placeholder="Email"
        />

        <FormField
          type="password"
          name="password"
          value={credentials.password}
          onChange={handleChange}
          placeholder="Contraseña"
          className="mt-4 border-2 rounded-lg "
        />
          <span className="mt-3 flex gap-2" onClick={()=> setIsChecked(!checked)} >Guardar La sesión?
          <input checked={checked} onChange={()=> setIsChecked(!checked)}  type="checkbox"/></span>


        <Button
          $variant="primary"
          type="submit"
          disabled={isDisabled}
          className="mt-4"
        >
          Iniciar Sesión
        </Button>
        {error ? null : email.length < 5 ? null : emailIsValid ? null : (
          <ErrorSpan children={"Por favor ingrese un email correcto"} />
        )}
        {error && (
          <ErrorSpan
            children={
              error.message === "Unauthorized"
                ? "Por favor ingrese un usuario y contraseña válidos"
                : error.message
            }
            onClick={() => setError(null)}
          />
        )}
      </form>
    </div>
  );
}

export default LoginPage;
