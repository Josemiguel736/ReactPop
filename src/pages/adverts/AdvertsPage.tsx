import { useEffect, useState } from "react";
import FilterAdverts from "../../components/filter/FilterAdverts";
import Adverts from "./Adverts";
import { AdvertType } from "./types";
import { getLastestAdverts } from "./service";
import { FilterContext } from "./context";
import { ApiClientError } from "../../api/error";
import { isApiClientError } from "../../api/client";
import Page501 from "../ErrorPages/501";
import LoadingPage from "../loadingPage/LoadingPage";


export default function AdvertsPage() {
  const [adverts, setAdverts] = useState<AdvertType[]>([]);
  const [error, setError] = useState<ApiClientError | null>(null);
  const [filteredAdverts, setFilteredAdverts] = useState<AdvertType[] | null>(
    null
  );
  const [isLoading, setIsLoading] = useState(true);


  useEffect(() => {
    getLastestAdverts()
      .then((response) => {
        setAdverts(response);
        setIsLoading(false)
      })
      .catch((error) => {
        setIsLoading(false)
        if (isApiClientError(error)) {
          setError(error);
          console.warn("ERROR IN API CALL TO ADVERTS FROM ADVERTS", error);
        } else if (error instanceof Error) {
          console.warn("GENERIC ERROR IN ADVERTS", error);
          return <Page501 error={error} />;
        }
      }
    );
  }, []);

  return isLoading ?  <LoadingPage/> :  error ? (
    <Page501 error={error} />
  ) : (
    <FilterContext.Provider value={{ filteredAdverts, setFilteredAdverts }}>
  <div className="flex flex-col lg:flex-row w-screen "> 
    <div className="flex justify-center w-full lg:w-1/5 mt-4 sm:mt-0 lg:ml-6  h-max  rounded-2xl p-4">
      <FilterAdverts adverts={adverts} />
    </div>
    
    <div className="flex justify-center items-center w-full flex-col ">
      <Adverts adverts={filteredAdverts ?? adverts} />
    </div>
  </div>
</FilterContext.Provider>
  );
}
