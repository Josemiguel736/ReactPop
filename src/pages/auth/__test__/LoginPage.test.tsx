import { render , screen } from "@testing-library/react"
import userEvent from "@testing-library/user-event"
import LoginPage from "../LoginPage"
import { Provider } from "react-redux"
import { AdvertType } from "../../adverts/types"
import * as actions from "../../../store/actions"
import { Credentials } from "../types"
import { State } from "../../../store/reducers"
import { Store } from "redux"


vi.mock("../../../store/actions")

describe("LoginPage", () => {

    const adverts:AdvertType[] =[
            {createdAt:"28/02/2025",
              id:"12/22as-22",
              name:"Test advert",
              photo:"photo/url",
              price:20,
              sale:true,
              tags:["motor"]        
            }]
    const tags:string[] = ["motor","lifestile","other tag"]

    const initialState:State = {
        auth:true,
        adverts:{data:adverts,loaded:true},
       tags:{data:tags,loaded:false,error:null},
        ui:{error:null,pending:false}    }

    const credentials:Credentials ={
        email:"admin@example.com",
        password:"admin"
    }

    const mockStore: Partial<Store<State>> = {
        getState: () => initialState,
        subscribe: vi.fn(),
        dispatch: vi.fn(),
      };

    const renderComponent = (error?:Error) =>{
        if(error){
            initialState.ui.error = error
        }
        return(
        render(<Provider store={mockStore as Store<State>}>
         <LoginPage/></Provider>))}

    test("should render", () =>{
       const {container} = renderComponent()
       expect(container).toMatchSnapshot()       
    })

    test("should dispatch login action", async ()=>{
        renderComponent()

       const emailInput = screen.getByPlaceholderText("Email")
       const passwordInput = screen.getByPlaceholderText("Contrase침a")
       const button = screen.getByRole("button")
       const checked = screen.getByRole("checkbox")

       expect(button).toHaveTextContent("Iniciar Sesi칩n")
       expect(button).toBeDisabled()
       
       await userEvent.type(emailInput,credentials.email)
       await userEvent.type(passwordInput,credentials.password)

       await userEvent.click(checked)

       expect(button).toBeEnabled()

       await userEvent.click(button)

       expect(actions.authLogin).toHaveBeenCalledWith(credentials,true)
       expect(actions.authLogin).toHaveBeenCalledTimes(1)
       
    })

    test("Test to render error", async ()=>{
        const {container} = renderComponent(new Error("UNAUTHORIZED"))        
        const alert = screen.getByRole("alert")

        expect(alert).toHaveTextContent("Por favor ingrese un usuario y contrase침a v치lidos")
        expect(container).toMatchSnapshot()

        await userEvent.click(alert)
        expect (actions.uiResetError).toHaveBeenCalled()      

    })

    test("Test to render other error", async ()=>{
        renderComponent(new Error("Test"))  
        const alert = screen.getByRole("alert")

        expect(alert).toHaveTextContent("Test")
        await userEvent.click(alert)

        expect (actions.uiResetError).toHaveBeenCalled()       
    })
})