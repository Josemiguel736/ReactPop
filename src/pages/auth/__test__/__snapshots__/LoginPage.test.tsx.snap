// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginPage > Test to render error 1`] = `
<div>
  <div
    class="h-1/3 mt-10 flex items-center justify-center"
  >
    <form
      class="text-white p-6 rounded-lg shadow-lg w-1/2 bg-sky-700 flex flex-col justify-center items-center text-center"
    >
      <h1
        class="text-2xl mt-2.5 mb-1"
      >
        Iniciar Sesión
      </h1>
      <div>
        <label>
          <input
            autocomplete="email"
            class="mt-4 border-2 rounded-lg "
            name="email"
            placeholder="Email"
            type="text"
            value=""
          />
        </label>
      </div>
      <div>
        <label>
          <input
            autocomplete="current-password"
            class="mt-4 border-2 rounded-lg "
            name="password"
            placeholder="Contraseña"
            type="password"
            value=""
          />
        </label>
      </div>
      <span
        class="mt-3 flex gap-2"
      >
        Guardar La sesión?
        <input
          type="checkbox"
        />
      </span>
      <button
        class="sc-FEMpB ihqouL mt-4"
        disabled=""
        type="submit"
      >
        Iniciar Sesión
      </button>
      <span
        class="text-red-500 bg-sky-950/50 pl-3.5 pr-3.5 "
        role="alert"
      >
        Por favor ingrese un usuario y contraseña válidos
      </span>
    </form>
  </div>
</div>
`;

exports[`LoginPage > should render 1`] = `
<div>
  <div
    class="h-1/3 mt-10 flex items-center justify-center"
  >
    <form
      class="text-white p-6 rounded-lg shadow-lg w-1/2 bg-sky-700 flex flex-col justify-center items-center text-center"
    >
      <h1
        class="text-2xl mt-2.5 mb-1"
      >
        Iniciar Sesión
      </h1>
      <div>
        <label>
          <input
            autocomplete="email"
            class="mt-4 border-2 rounded-lg "
            name="email"
            placeholder="Email"
            type="text"
            value=""
          />
        </label>
      </div>
      <div>
        <label>
          <input
            autocomplete="current-password"
            class="mt-4 border-2 rounded-lg "
            name="password"
            placeholder="Contraseña"
            type="password"
            value=""
          />
        </label>
      </div>
      <span
        class="mt-3 flex gap-2"
      >
        Guardar La sesión?
        <input
          type="checkbox"
        />
      </span>
      <button
        class="sc-FEMpB ihqouL mt-4"
        disabled=""
        type="submit"
      >
        Iniciar Sesión
      </button>
    </form>
  </div>
</div>
`;
